private/bufpkg/bufmodule/module_file_set.go

```diff
@@ -57,10 +57,27 @@ func newModuleFileSet(
 			),
 		)
 	}
+	multiReadBucket := newMultiModuleReadBucket(moduleReadBuckets...)
+	seenPaths := map[string][]string{}
+	if err := multiReadBucket.WalkModuleFiles(context.Background(), "", func(moduleObjectInfo *moduleObjectInfo) error {
```

Comment:
This feels *really* expensive to do on construction of a `ModuleFileSet` every time - this could be 1000's of files just to change an error message

Reply:
Yeah, that's very fair... the other option that @amckinney and I explored is extending the `ObjectInfo` interface to take in some metadata and return a "fully-qualified path" (in addition to `Path()` and `ExternalPath()`), so we don't affect the output of the existing implementations. That is probably the other reasonable route.

Reply:
Another way to think of the `ObjectInfo` extension is supporting another `interface` that could be up-casted to, but it's really the same thing e.g.

```go
 // I don't know what to call this right now.
type SomethingObjectInfo interface {
  ObjectInfo

  // FullyQualifiedPath returns a string that contains the fully-qualified path
  // for this object. If set, this will defer from the ExternalPath representation.
  FullyQualifiedPath() string
}
```

Then you could check for its existence:

```go
if v, ok := objectInfo.(SomethingObjectInfo); ok {
  ...
}
```

This isn't great though since it effects the `storage` APIs and it's unclear how this would be used by other `storage` implementations since we already have `ExternalPath` which _should_ represent what `FullyQualifiedPath` does in this specific case.

As for validating this for every `ModuleFileSet`, this operation is already happening: it's just being brought up a layer. We'll end up iterating over each file one more time than we previously did, but it's still a O(n) operation. Given that a ModuleFileSet should really only be in the ~1000s of files, this probably isn't even noticeable in practice - we're just walking over each file. It also feels natural for the `ModuleFileSet` to validate itself this way. With this, we can make other guarantees about the validity of the `ModuleFileSet`, which could be (potentially) leveraged elsewhere.

I could be wrong w.r.t the cost here, but that's my high-level take.

Reply:
With all that said, there _are_ implications around this cost in workspace environments. GIven that we currently build a different `ModuleFileSet` for each workspace directory, there's a lot of duplicative effort here. This ties back into some of the thoughts I have in the summary of #544, however.